src/canister/individual_user_template/src/api/canister_lifecycle/post_upgrade.rs

```diff
@@ -23,18 +23,36 @@ fn post_upgrade() {
 
 fn restore_data_from_stable_memory() {
     let heap_data = memory::get_upgrades_memory();
+    let mut upgrade_reader = Reader::new(&heap_data, 0);
+
     let mut heap_data_len_bytes = [0; 4];
-    heap_data.read(0, &mut heap_data_len_bytes);
+    upgrade_reader.read(&mut heap_data_len_bytes).unwrap();
     let heap_data_len = u32::from_le_bytes(heap_data_len_bytes) as usize;
 
     let mut canister_data_bytes = vec![0; heap_data_len];
-    heap_data.read(4, &mut canister_data_bytes);
+    upgrade_reader.read(&mut canister_data_bytes).unwrap();
 
     let canister_data =
         de::from_reader(&*canister_data_bytes).expect("Failed to deserialize heap data");
     CANISTER_DATA.with(|canister_data_ref_cell| {
         *canister_data_ref_cell.borrow_mut() = canister_data;
     });
+
+    upgrade_reader.read(&mut heap_data_len_bytes).unwrap();
+
+    let mut pump_n_dump_data_bytes = vec![0; heap_data_len];
```

Comment:
Shouldn't we have to assign `heap_data_len` to new value here?

```suggestion
    upgrade_reader.read(&mut heap_data_len_bytes).unwrap();
    heap_data_len = u32::from_le_bytes(heap_data_len_bytes);
    let mut pump_n_dump_data_bytes = vec![0; heap_data_len];
```

Reply:
Great catch, i didn't notice at all